kitten-scientists.js

```diff
@@ -441,17 +439,59 @@ TradeManager.prototype = {
         message('Trade: ' + amount + 'x ' + race.title);
     },
     getLowestTradeAmount: function (name) {
-        var amount = 0;
+        var amount = -1;
+        var highestCapacity = -1;
         var consume = options.consume;
         var materials = this.getMaterials(name);
+        var race = this.getRace(name);
 
         for (var i in materials) {
             var total = this.craftManager.getValueAvailable(i) * consume / materials[i];
 
-            amount = (0 === amount || total < amount) ? total : amount;
+            amount = (-1 === amount || total < amount) ? total : amount;
         }
 
-        return amount;
+        // Loop through the items obtained by the race, and determine
+        // which good has the most space left. Once we've determined this,
+        // reduce the amount by this capacity. This ensures that we continue to trade
+        // as long as at least one resource has capacity, and we never over-trade.
+        for (var s in race.sells) {
+            var resource = this.craftManager.getResource(i.name);
+            var max = 0;
+
+            // No need to process resources that don't cap
+            if (!resource.maxValue) continue;
+
+            // Zebras special cased titanium taken directly from game code
+            if (race.name == "zebras" && s.name == "titanium") {
+				var val = 1.5 + (1.5 * this.game.resPool.get("ship").value / 100 * 2);
+				max = Math.ceil(val);
+			} else {
+			    var sratio = s.seasons[game.calendar.getCurSeason().name];
+				var tratio = self.game.bld.getEffect("tradeRatio");
+				var val = s.value + s.value * tratio;
+
+				max = val * sratio * (1 + s.delta/2);
+			}
+
+			capacity = (resource.maxValue - resource.value) / max;
```

Comment:
I like how this is doing its checks, nice and simple, but the formatting needs to be updated to make it easier to consume.